share/github-backup-utils/ghe-backup-storage

```diff
@@ -0,0 +1,129 @@
+#!/usr/bin/env bash
+#/ Usage: ghe-backup-storage
+#/ Take an online, incremental snapshot of all Alambic Storage data using the
+#/ calculated routes method.
+#/
+#/ Note: This command typically isn't called directly. It's invoked by
+#/ ghe-backup.
+set -e
+
+# Bring in the backup configuration
+. $( dirname "${BASH_SOURCE[0]}" )/ghe-backup-config
+
+bm_start "$(basename $0)"
+
+# Set up remote host and root backup snapshot directory based on config
+host="$GHE_HOSTNAME"
+backup_dir="$GHE_SNAPSHOT_DIR/storage"
+
+# Verify rsync is available.
+if ! rsync --version 1>/dev/null 2>&1; then
+    echo "Error: rsync not found." 1>&2
+    exit 1
+fi
+
+# Perform a host-check and establish GHE_REMOTE_XXX variables.
+ghe_remote_version_required "$host"
+
+# Split host:port into parts
+port=$(ssh_port_part "$GHE_HOSTNAME")
+host=$(ssh_host_part "$GHE_HOSTNAME")
+
+# Add user / -l option
+user="${host%@*}"
+[ "$user" = "$host" ] && user="admin"
+
+hostnames=$host
+ssh_config_file_opt=
+tempdir=$(mktemp -d -t backup-utils-restore-XXXXXX)
+opts="$GHE_EXTRA_SSH_OPTS"
+
+# storage server hostnames under cluster
+if [ "$GHE_BACKUP_STRATEGY" = "cluster" ]; then
+  ssh_config_file="$tempdir/ssh_config"
+  ssh_config_file_opt="-F $ssh_config_file"
+  opts="$opts -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o PasswordAuthentication=no"
+  hostnames=$(ghe-cluster-nodes "$GHE_HOSTNAME" "storage-server")
+  ghe-ssh-config "$GHE_HOSTNAME" "$hostnames" > "$ssh_config_file"
+fi
+
+# Make sure root backup dir exists if this is the first run
+mkdir -p "$backup_dir"
+
+# Removes the remote sync-in-progress file on exit, re-enabling GC operations
+# on the remote instance.
+cleanup() {
+  # Enable remote maintenance operations
+  for hostname in $hostnames; do
+    ghe-gc-enable $ssh_config_file_opt $hostname:$port
```

Comment:
if ssh fails on any of these the rest of the operations will fail? Maybe echo the failure to enable gc operations on the specified host?

Reply:
Good idea. I'll do this elsewhere too.